synapse/storage/data_stores/main/ui_auth.py

```diff
@@ -0,0 +1,253 @@
+                retcols=("clientdict", "uri", "method", "description"),
+                desc="get_session",
+            )
+        except StoreError:
+            raise SynapseError(400, "Unknown session ID: %s" % session_id)
+
+        result["clientdict"] = json.loads(result["clientdict"])
+
+        return result
+
+    def delete_old_sessions(self, expiration_time: int):
+        """
+        Remove sessions which were last used earlier than the expiration time.
+
+        Args:
+            expiration_time: The latest time that is still considered valid.
+                This is an epoch time in milliseconds.
+
+        """
+        return self.db.runInteraction(
+            "delete_sessions", self._delete_old_sessions, expiration_time
+        )
+
+    def _delete_old_sessions(self, txn, expiration_time: int):
+        # Get the expired sessions.
+        sql = "SELECT session_id FROM ui_auth_sessions WHERE last_used <= ?"
+        txn.execute(sql, [expiration_time])
+        session_ids = [r[0] for r in txn.fetchall()]
+
+        # Delete the corresponding completed credentials.
+        self.db.simple_delete_many_txn(
+            txn,
+            table="ui_auth_sessions_credentials",
+            column="session_id",
+            iterable=session_ids,
+            keyvalues={},
+        )
+
+        # Finally, delete the sessions.
+        self.db.simple_delete_many_txn(
+            txn,
+            table="ui_auth_sessions",
+            column="session_id",
+            iterable=session_ids,
+            keyvalues={},
+        )
+
+    async def mark_stage_complete(
+        self,
+        session_id: str,
+        stage_type: str,
+        identity: Union[str, bool, Dict[str, any]],
+    ):
+        """
+        Mark a session stage as completed.
+
+        Args:
+            session_id: The ID of the corresponding session.
+            stage_type: The completed stage type.
+            identity: The identity authenticated by the stage.
+        """
+        await self.db.runInteraction(
+            "mark_stage_complete",
+            self._mark_stage_completed,
+            session_id,
+            stage_type,
+            identity,
+        )
+
+    def _mark_stage_completed(
+        self,
+        txn,
+        session_id: str,
+        stage_type: str,
+        identity: Union[str, bool, Dict[str, any]],
+    ):
+        # Add (or update) the results of the current stage to the database.
+        self.db.simple_upsert_txn(
+            txn,
+            table="ui_auth_sessions_credentials",
+            keyvalues={"session_id": session_id, "stage_type": stage_type},
+            values={"identity": json.dumps(identity)},
+        )
+        # Mark the session as still in use.
+        self.db.simple_update_one_txn(
+            txn,
+            table="ui_auth_sessions",
+            keyvalues={"session_id": session_id},
+            updatevalues={"last_used": self.hs.get_clock().time_msec()},
+        )
```

Comment:
I do wonder if updating `last_used` here is really useful and if we can just track creation time. The delay for expiring these is 48 hours, but I'd expect the entire session to last only a few minutes in most cases.

Reply:
👍  This simplifies things a bit. Great!